### PR TITLE
Refer to "associated functions" instead of "static methods"

### LINKED ISSUE #59782: E0424 The self keyword was used in a static method.
E0424 states:

> The self keyword was used in a `static` method.

But `static` method is probably wrong, it should just say a non-method (I don't think we have a technical term to refer to methods that do not take an arbitrary self type as first argument). 

It continues with:

> Please check if the method's argument list should have contained self, &self, or &mut self (in case you didn't want to create a static method), and add it if so.

That's outdated, `&mut Pin<Self>` (and others) work too. So this should be updated to cover arbitrary `Self` types.

### PR DESCRIPTION
**Pull request by estebank**
Fix #59782.

### CODE REVIEW

**Comment by estebank** [comment]
cc @Centril @petrochenkov

File: src/librustc_resolve/error_codes.rs [hunk 1]
```diff
@@ -1013,33 +1013,39 @@ fn h1() -> i32 {
 "##,
 
 E0424: r##"
-The `self` keyword was used in a static method.
+The `self` keyword was used inside of an associated function instead of inside
+of a method. Associated functions have no "`self` receiver" argument, and are
+equivalent to regular functions which exist in the namespace of a trait.
```
Comment (Centril) [inline src/librustc_resolve/error_codes.rs]: This comment re. equivalence is rather strong and perhaps unnecessary?

File: src/librustc_resolve/error_codes.rs [hunk 2]
```diff
@@ -1013,33 +1013,39 @@ fn h1() -> i32 {
...
+Methods, on the other hand, have a `self` reciver argument, like `self`,
```
Comment (Centril) [inline src/librustc_resolve/error_codes.rs]:
```suggestion
Methods, on the other hand, have a `self` receiver parameter, like `self`,
```

File: src/libsyntax_ext/deriving/generic/mod.rs [hunk 3]
```diff
@@ -1055,9 +1055,7 @@ impl<'a> MethodDef<'a> {
                 })
                 .collect()
         } else {
-            cx.span_bug(trait_.span,
-                        "no self arguments to non-static method in generic \
-                         `derive`")
+            cx.span_bug(trait_.span, "no self arguments for method in generic `derive`")
```
Comment (Centril) [inline src/libsyntax_ext/deriving/generic/mod.rs]:
```suggestion
            cx.span_bug(trait_.span, "no self parameters for method in generic `derive`")
```

File: src/librustc_resolve/error_codes.rs [hunk 4]
```diff
@@ -1013,33 +1013,39 @@ fn h1() -> i32 {
...
 
 impl Foo {
-    fn bar(self) {}
+    // `bar` is a method, because it has a receiver argument.
+    fn bar(&self) {}
 
+    // `foo` is an associated function, because it has no receiver argument.
     fn foo() {
-        self.bar(); // error: `self` is not available in a static method.
+        self.bar(); // error: `self` is not available in an associated function
```
[the remark this answers is not in the document]
Comment (estebank) [reply src/librustc_resolve/error_codes.rs]: Something like "all functions inside a trait are associated functions, associated functions with a self receiver are methods, only methods can refer to self in their bodies"?

[the remark this answers is not in the document]
Comment (Centril) [reply src/librustc_resolve/error_codes.rs]: Associated functions are also those in impls (inherent or otherwise).

**Comment by estebank** [comment]
@Centril can you take a look at the new wording?

File: src/librustc_resolve/error_codes.rs [hunk 5]
```diff
@@ -1013,33 +1013,40 @@ fn h1() -> i32 {
 "##,
 
 E0424: r##"
-The `self` keyword was used in a static method.
+The `self` keyword was used inside of an associated function without a "`self`
+receiver" parameter. The `self` keyword can only be used inside methods, which
```
Comment (GuillaumeGomez) [inline src/librustc_resolve/error_codes.rs]: Please put the explananation **after** the failing example.

Comment (estebank) [reply src/librustc_resolve/error_codes.rs]: Done.

Comment (estebank) [reply src/librustc_resolve/error_codes.rs]: Done.

**Pull request merged**